echopype/preprocess/api.py

```diff
@@ -329,3 +284,427 @@ def remove_noise(ds_Sv, ping_num, range_sample_num, noise_max=None, SNR_threshol
 
 def regrid():
     return 1
+
+
+def get_bin_indices(
+    echo_range: np.ndarray, bins_er: np.ndarray, times: np.ndarray, bins_time: np.ndarray
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    Obtains the bin index of ``echo_range`` and ``times`` based
+    on the binning ``bins_er`` and ``bins_time``, respectively.
+
+    Parameters
+    ----------
+    echo_range: np.ndarray
+        2D array of echo range values
+    bins_er: np.ndarray
+        1D array (used by np.digitize) representing the binning required for ``echo_range``
+    times: np.ndarray
+        1D array corresponding to the time values that should be binned
+    bins_time: np.ndarray
+        1D array (used by np.digitize) representing the binning required for ``times``
+
+    Returns
+    -------
+    digitized_echo_range: np.ndarray
+        2D array of bin indices for ``echo_range``
+    bin_time_ind: np.ndarray
+        1D array of bin indices for ``times``
+    """
+
+    # get bin index for each echo range value
+    digitized_echo_range = np.digitize(echo_range, bins_er, right=False)
+
+    # turn datetime into integers, so we can use np.digitize
+    if isinstance(times, dask.array.Array):
+        times_i8 = times.compute().data.view("i8")
+    else:
+        times_i8 = times.view("i8")
+
+    # turn datetime into integers, so we can use np.digitize
+    bins_time_i8 = bins_time.view("i8")
+
+    # get bin index for each time
+    bin_time_ind = np.digitize(times_i8, bins_time_i8, right=False)
+
+    return digitized_echo_range, bin_time_ind
+
+
+def bin_and_mean_echo_range(
+    arr: Union[np.ndarray, dask.array.Array], digitized_echo_range: np.ndarray, n_bin_er: int
+) -> Union[np.ndarray, dask.array.Array]:
+    """
+    Bins and means ``arr`` with respect to the ``echo_range`` bins.
+
+    Parameters
+    ----------
+    arr: np.ndarray or dask.array.Array
+        2D array to bin and mean
+    digitized_echo_range: np.ndarray
+        2D array of bin indices for ``echo_range``
+    n_bin_er: int
+        The number of echo range bins
+
+    Returns
+    -------
+    er_means: np.ndarray or dask.array.Array
+        2D array representing the bin and mean of ``arr`` along ``echo_range``
+    """
+
+    binned_means = []
+    for bin_er in range(1, n_bin_er):
+
+        # bin and mean echo_range dimension
+        er_selected_data = np.nanmean(arr[:, digitized_echo_range == bin_er], axis=1)
+
+        # collect all echo_range bins
+        binned_means.append(er_selected_data)
```

Comment:
If we wrap this line
```
er_selected_data = np.nanmean(arr[:, digitized_echo_range == bin_er], axis=1)
```
into a function and delay it, wouldn't dask be able to distributing the nanmean computation instead of computing each binned averages in a for loop?

Reply:
If the input to `compute_MVBS` is a lazy Dataset, then in this code section `arr` will also be delayed. For this reason, there is no computation being done in the for loop you highlighted and the output of this function will also be delayed. When we actually compute this, Dask will efficiently distribute the computation. If we wrapped the nanmean in a delayed function then this would lead to inefficient computation and overwhelm the scheduler with too many tasks.

Reply:
> there is no computation being done in the for loop you highlighted and the output of this function will also be delayed

Yes I understood this, my question was whether this was computed sequentially or simultaneously (distributed).

Since this is a for loop, seems that Dask would distribute within each iteration of the loop, and not running all the different bins in parallel. When you say
```
When we actually compute this, Dask will efficiently distribute the computation.
```
what is being distributed efficiently?

> If we wrapped the nanmean in a delayed function then this would lead to inefficient computation and overwhelm the scheduler with too many tasks.

Do you mean that the number of tasks is the cause of inefficiency? Or the number of small tasks, whose size is related to the size of the bins along echo_range and ping_time?